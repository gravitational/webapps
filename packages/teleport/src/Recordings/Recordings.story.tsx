/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { makeEvent } from 'teleport/services/audit';
import { Router } from 'react-router';
import { createMemoryHistory } from 'history';
import { Context, ContextProvider } from 'teleport';
import Recordings from './Recordings';

export default {
  title: 'Teleport/Recordings',
};

export const Loaded = () => {
  const ctx = new Context();
  ctx.auditService.fetchEvents = () =>
    Promise.resolve({
      overflow: false,
      events: events.map(makeEvent),
    });

  return render(ctx);
};

export const Overflow = () => {
  const ctx = new Context();
  ctx.auditService.fetchEvents = () =>
    Promise.resolve({
      overflow: true,
      events: [],
    });

  return render(ctx);
};

export const Processing = () => {
  const ctx = new Context();
  ctx.auditService.fetchEvents = () => new Promise(() => null);
  return render(ctx);
};

export const Failed = () => {
  const ctx = new Context();
  ctx.auditService.fetchEvents = () =>
    Promise.reject(new Error('server error'));
  return render(ctx);
};

function render(ctx) {
  const history = createMemoryHistory({
    initialEntries: ['/web/cluster/localhost/audit/events'],
    initialIndex: 0,
  });

  return (
    <ContextProvider ctx={ctx}>
      <Router history={history}>
        <Recordings />
      </Router>
    </ContextProvider>
  );
}

const events = [
  {
    code: 'T2004I',
    ei: 10,
    event: 'session.end',
    namespace: 'default',
    sid: '426485-6491-11e9-80a1-427cfde50f5a',
    time: '2019-04-22T00:00:51.543Z',
    uid: '6bf836ee-197c-453e-98e5-31511935f22a',
    user: 'admin@example.com',
    participants: ['one', 'two'],
    server_id: 'serverId',
    server_hostname: 'apple-node',
    interactive: true,
  },
  {
    code: 'T2004I',
    ei: 10,
    event: 'session.end',
    namespace: 'default',
    sid: '377875-6491-11e9-80a1-427cfde50f5a',
    time: '2019-04-22T00:00:51.543Z',
    uid: '6bf836ee-197c-453e-98e5-31511935f22a',
    user: 'admin@example.com',
    participants: ['one', 'two'],
    server_id: 'serverId',
    server_hostname: 'peach-node',
    interactive: true,
  },
  {
    code: 'T2004I',
    ei: 10,
    event: 'session.end',
    namespace: 'default',
    sid: '1682c475-04fd-5c4f-8881-7d4d44a3bfb1',
    time: '2019-05-22T00:00:51.543Z',
    uid: '6bf836ee-197c-453e-98e5-31511935f22a',
    user: 'admin@example.com',
    participants: ['one', 'two'],
    server_id: 'serverId',
    server_hostname: 'pear-node',
    interactive: true,
  },
];
