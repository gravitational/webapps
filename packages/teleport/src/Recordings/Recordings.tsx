/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import {
  FeatureBox,
  FeatureHeader,
  FeatureHeaderTitle,
} from 'teleport/components/Layout';
import RecordList from './RecordList';
import RangePicker from 'teleport/components/EventRangePicker';
import { Danger } from 'design/Alert';
import { Flex, Indicator, Box } from 'design';
import InputSearch from 'teleport/components/InputSearch';
import useTeleport from 'teleport/useTeleport';
import useAuditEvents from 'teleport/useAuditEvents';
import useStickyClusterId from 'teleport/useStickyClusterId';
import { eventCodes } from 'teleport/services/audit';

export default function Container() {
  const teleCtx = useTeleport();
  const { clusterId } = useStickyClusterId();
  const state = useAuditEvents(teleCtx, clusterId, eventCodes.SESSION_END);
  return <Recordings {...state} />;
}

export function Recordings(props: ReturnType<typeof useAuditEvents>) {
  const {
    attempt,
    range,
    rangeOptions,
    setRange,
    events,
    searchValue,
    clusterId,
    setSearchValue,
    fetchMore,
    fetchStatus,
  } = props;

  return (
    <FeatureBox>
      <FeatureHeader alignItems="center">
        <FeatureHeaderTitle mr="8">Session Recordings</FeatureHeaderTitle>
        <RangePicker
          ml="auto"
          range={range}
          ranges={rangeOptions}
          onChangeRange={setRange}
        />
      </FeatureHeader>
      <Flex
        mb={4}
        alignItems="center"
        flex="0 0 auto"
        justifyContent="flex-start"
      >
        <InputSearch mr="3" onChange={setSearchValue} />
      </Flex>
      {attempt.status === 'failed' && <Danger> {attempt.statusText} </Danger>}
      {attempt.status === 'processing' && (
        <Box textAlign="center" m={10}>
          <Indicator />
        </Box>
      )}
      {attempt.status === 'success' && (
        <RecordList
          searchValue={searchValue}
          events={events}
          clusterId={clusterId}
          pageSize={50}
          fetchMore={fetchMore}
          fetchStatus={fetchStatus}
        />
      )}
    </FeatureBox>
  );
}
