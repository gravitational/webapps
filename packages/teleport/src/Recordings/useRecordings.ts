import { useState, useEffect, useMemo } from 'react';
import useAttempt from 'shared/hooks/useAttemptNext';
import Ctx from 'teleport/teleportContext';
import useStickyClusterId from 'teleport/useStickyClusterId';
import {
  getRangeOptions,
  EventRange,
} from 'teleport/components/EventRangePicker';
import { Recording } from 'teleport/services/recordings';

export default function useRecordings(ctx: Ctx) {
  const { clusterId } = useStickyClusterId();
  const rangeOptions = useMemo(() => getRangeOptions(), []);
  const [searchValue, setSearchValue] = useState('');
  const [range, setRange] = useState<EventRange>(rangeOptions[0]);
  const { attempt, setAttempt, run } = useAttempt('processing');
  const [results, setResults] = useState<RecordingsResult>({
    recordings: [],
    fetchStartKey: '',
    fetchStatus: '',
  });

  function fetchMore() {
    setResults({
      ...results,
      fetchStatus: 'loading',
    });
    ctx.recordingsService
      .fetchRecordings(clusterId, {
        ...range,
        startKey: results.fetchStartKey,
      })
      .then(res =>
        setResults({
          recordings: [...results.recordings, ...res.recordings],
          fetchStartKey: res.startKey,
          fetchStatus: res.startKey ? '' : 'disabled',
        })
      )
      .catch((err: Error) => {
        setAttempt({ status: 'failed', statusText: err.message });
      });
  }

  function fetch() {
    run(() =>
      ctx.recordingsService
        .fetchRecordings(clusterId, {
          ...range,
        })
        .then(res =>
          setResults({
            recordings: res.recordings,
            fetchStartKey: res.startKey,
            fetchStatus: res.startKey ? '' : 'disabled',
          })
        )
    );
  }

  useEffect(() => {
    fetch();
  }, [clusterId, range]);

  return {
    ...results,
    attempt,
    range,
    rangeOptions,
    setRange,
    searchValue,
    clusterId,
    setSearchValue,
    fetchMore,
  };
}

type RecordingsResult = {
  recordings: Recording[];
  fetchStatus: 'loading' | 'disabled' | '';
  fetchStartKey: string;
};

export type State = ReturnType<typeof useRecordings>;
