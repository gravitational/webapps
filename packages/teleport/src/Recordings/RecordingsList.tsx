/*
Copyright 2019 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from 'react';
import { ButtonBorder } from 'design';
import Table, { Cell, TextCell } from 'design/DataTable';
import { dateTimeMatcher } from 'design/utils/match';
import { displayDateTime } from 'shared/services/loc';
import cfg from 'teleport/config';
import { Recording, RecordingType } from 'teleport/services/recordings';
import { State } from './useRecordings';
import Icon, * as Icons from 'design/Icon/Icon';

export default function RecordingsList(props: Props) {
  const {
    recordings = [],
    clusterId,
    pageSize = 50,
    fetchMore,
    fetchStatus,
  } = props;

  return (
    <Table
      data={recordings}
      columns={[
        {
          headerText: 'Type',
          key: 'recordingType',
          isSortable: true,
          render: ({ recordingType }) => renderIconCell(recordingType),
        },
        {
          key: 'hostname',
          headerText: 'Name',
          isSortable: true,
        },
        {
          key: 'users',
          headerText: 'User(s)',
          render: ({ users }) => (
            <Cell style={{ wordBreak: 'break-word' }}>{users}</Cell>
          ),
          isSortable: true,
        },
        {
          key: 'duration',
          headerText: 'Duration',
          isSortable: true,
          render: ({ durationText }) => <TextCell data={durationText} />,
        },
        {
          key: 'createdDate',
          headerText: 'Created',
          isSortable: true,
          render: ({ createdDate }) => (
            <Cell>{displayDateTime(createdDate)}</Cell>
          ),
        },
        {
          key: 'sid',
          headerText: 'Session ID',
        },
        {
          altKey: 'play-btn',
          render: recording => renderPlayCell(recording, clusterId),
        },
      ]}
      emptyText="No Recordings Found"
      pagination={{ pageSize }}
      fetching={{ onFetchMore: fetchMore, fetchStatus }}
      initialSort={{
        key: 'createdDate',
        dir: 'DESC',
      }}
      isSearchable
      searchableProps={[
        'recordingType',
        'hostname',
        'description',
        'createdDate',
        'sid',
        'users',
        'durationText',
      ]}
      customSearchMatchers={[dateTimeMatcher(['createdDate'])]}
    />
  );
}

const renderIconCell = (type: RecordingType) => {
  let icon = Icons.Cli;
  if (type === 'desktop') {
    icon = Icons.Desktop;
  } else if (type === 'k8s') {
    icon = Icons.Kubernetes;
  }

  return (
    <Cell>
      <Icon p={1} mr={3} fontSize={3} as={icon} />
    </Cell>
  );
};

const renderPlayCell = (
  { description, sid, recordingType, playable, duration }: Recording,
  clusterId: string
) => {
  if (!playable) {
    return (
      <Cell align="right" style={{ color: '#9F9F9F' }}>
        {description}
      </Cell>
    );
  }

  const url = cfg.getPlayerRoute(
    { clusterId, sid },
    {
      recordingType,
      durationMs: recordingType === 'desktop' ? duration : undefined,
    }
  );
  return (
    <Cell align="right">
      <ButtonBorder
        kind="primary"
        as="a"
        href={url}
        width="80px"
        target="_blank"
        size="small"
      >
        Play
      </ButtonBorder>
    </Cell>
  );
};

type Props = {
  pageSize?: number;
  recordings: State['recordings'];
  clusterId: State['clusterId'];
  fetchMore: State['fetchMore'];
  fetchStatus: State['fetchStatus'];
};
