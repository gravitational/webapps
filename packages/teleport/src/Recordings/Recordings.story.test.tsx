import React from 'react';
import { Loaded } from './Recordings.story';
import { render, waitFor } from 'design/utils/testing';

test('rendering of Session Recordings', async () => {
  const { container } = render(<Loaded />);

  await waitFor(() => document.querySelector('table'));
  expect(container).toMatchSnapshot();
});
