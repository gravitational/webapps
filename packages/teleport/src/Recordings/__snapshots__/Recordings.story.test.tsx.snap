// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering of Session Recordings 1`] = `
.c14 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  font-size: 10px;
  min-height: 24px;
  padding: 0px 16px;
  width: 80px;
}

.c14:active {
  opacity: 0.56;
}

.c14:hover,
.c14:focus {
  background: #651FFF;
}

.c14:active {
  background: #354AA4;
}

.c14:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c5 {
  display: inline-block;
  transition: color .3s;
  margin-left: 8px;
  margin-right: -8px;
  color: rgba(255,255,255,0.87);
  font-size: 14px;
}

.c11 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
  font-size: 16px;
}

.c13 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
}

.c3 {
  line-height: 1.5;
  border-radius: 4px;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  border: 1px solid;
  box-sizing: border-box;
  background-color: transparent;
  cursor: pointer;
  font-family: inherit;
  font-weight: bold;
  outline: none;
  opacity: .56;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all .3s;
  -webkit-font-smoothing: antialiased;
  border-color: rgba(255,255,255,0.87);
  color: #FFFFFF;
  font-size: 10px;
  min-height: 24px;
  padding: 0px 16px;
  margin-left: auto;
  width: auto;
  border-color: rgba(255,255,255,0.87);
  color: rgba(255,255,255,0.87);
}

.c3:hover {
  opacity: 1;
}

.c3:active {
  opacity: .24;
}

.c3 > span {
  display: flex;
  align-items: center;
  justify-content: center;
}

.c3:hover,
.c3:focus {
  border-color: #FFFFFF;
  color: #FFFFFF;
}

.c3:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c3:hover,
.c3:focus {
  border-color: #FFFFFF;
  color: #FFFFFF;
}

.c9 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 12px;
  line-height: 16px;
  margin: 0px;
  color: #FFFFFF;
}

.c6 {
  box-sizing: border-box;
  margin-bottom: 24px;
  flex: 0 0 auto;
  display: flex;
  align-items: center;
  justify-content: flex-start;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  margin-right: 56px;
  white-space: nowrap;
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  display: flex;
  overflow: auto;
  width: 100%;
  height: 100%;
  flex-direction: column;
}

.c0::after {
  content: ' ';
  padding-bottom: 24px;
}

.c12 {
  background: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-collapse: collapse;
  border-spacing: 0;
  border-radius: 8px;
  font-size: 12px;
  width: 100%;
}

.c12 > thead > tr > th,
.c12 > tbody > tr > th,
.c12 > tfoot > tr > th,
.c12 > thead > tr > td,
.c12 > tbody > tr > td,
.c12 > tfoot > tr > td {
  padding: 8px 8px;
  vertical-align: middle;
}

.c12 > thead > tr > th:first-child,
.c12 > tbody > tr > th:first-child,
.c12 > tfoot > tr > th:first-child,
.c12 > thead > tr > td:first-child,
.c12 > tbody > tr > td:first-child,
.c12 > tfoot > tr > td:first-child {
  padding-left: 24px;
}

.c12 > thead > tr > th:last-child,
.c12 > tbody > tr > th:last-child,
.c12 > tfoot > tr > th:last-child,
.c12 > thead > tr > td:last-child,
.c12 > tbody > tr > td:last-child,
.c12 > tfoot > tr > td:last-child {
  padding-right: 24px;
}

.c12 > thead > tr > th {
  background: #111B48;
  color: #FFFFFF;
  cursor: pointer;
  font-size: 10px;
  font-weight: 400;
  padding-bottom: 0;
  padding-top: 0;
  text-align: left;
  opacity: 0.75;
  text-transform: uppercase;
  white-space: nowrap;
}

.c12 > thead > tr > th .c4 {
  font-weight: bold;
  font-size: 8px;
  margin-left: 8px;
}

.c12 > tbody > tr > td {
  color: rgba(255,255,255,0.87);
  line-height: 16px;
}

.c12 tbody tr {
  border-bottom: 1px solid #1C254D;
}

.c12 tbody tr:hover {
  background-color: rgb(37,49,98);
}

.c12 tbody tr:hover:last-child td:first-child {
  border-bottom-left-radius: 8px;
}

.c12 tbody tr:hover:last-child td:last-child {
  border-bottom-right-radius: 8px;
}

.c10 {
  box-sizing: border-box;
  display: flex;
}

.c10 button {
  background: none;
  border: none;
  border-radius: 200px;
  cursor: pointer;
  padding: 0;
  margin: 0 0 0 8px;
  outline: none;
  transition: all 0.3s;
  text-align: center;
}

.c10 button:hover,
.c10 button:focus {
  background: #1C254D;
}

.c10 button:hover .c4,
.c10 button:focus .c4 {
  opacity: 1;
}

.c10 button .c4 {
  opacity: 0.4;
  font-size: 20px;
  transition: all 0.3s;
}

.c8 {
  padding: 8px 24px;
  display: flex;
  height: 24px;
  flex-shrink: 0;
  align-items: center;
  justify-content: space-between;
  background: #222C59;
  border-top-left-radius: 8px;
  border-top-right-radius: 8px;
}

.c7 {
  box-sizing: border-box;
  font-size: 12px;
  min-width: 200px;
  outline: none;
  border: none;
  border-radius: 200px;
  height: 32px;
  transition: all .2s;
  background: #222C59;
  margin-right: 16px;
  padding-left: 16px;
  padding-right: 16px;
  color: rgba(255,255,255,0.87);
}

.c7:hover {
  background: #2C3A73;
}

.c7:focus,
.c7:active {
  background: #2C3A73;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  color: rgba(255,255,255,0.87);
}

.c7::placeholder {
  color: rgba(255,255,255,0.24);
  font-size: 12px;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
       Session Recordings
    </div>
    <button
      class="c3"
      style="min-width: 100px;"
      width="auto"
    >
      <span>
        Today
        <span
          class="c4 c5 icon icon-caret-down "
          color="text.primary"
          font-size="2"
        />
      </span>
    </button>
  </div>
  <div
    class="c6"
  >
    <input
      class="c7"
      color="text.primary"
      placeholder="SEARCH..."
      value=""
    />
  </div>
  <div>
    <nav
      class="c8"
    >
      <div
        class="c9"
        color="primary.contrastText"
      >
        SHOWING 
        <strong>
          1
        </strong>
         - 
        <strong>
          3
        </strong>
         
        of 
        <strong>
          3
        </strong>
      </div>
      <div
        class="c10"
      >
        <button
          disabled=""
          title="Previous Page"
        >
          <span
            class="c4 c11 icon icon-arrow-left-circle "
            color="light"
            font-size="3"
          />
        </button>
        <button
          disabled=""
          title="Next Page"
        >
          <span
            class="c4 c11 icon icon-arrow-right-circle "
            color="light"
            font-size="3"
          />
        </button>
      </div>
    </nav>
    <table
      class="c12"
    >
      <thead>
        <tr>
          <th>
            Node
          </th>
          <th>
            User(s)
          </th>
          <th>
            <a>
              Duration
              <span
                class="c4 c13 icon icon-chevrons-expand-vertical "
                color="light"
              />
            </a>
          </th>
          <th>
            <a>
              Created
              <span
                class="c4 c13 icon icon-chevron-up "
                color="light"
              />
            </a>
          </th>
          <th>
            Session ID
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            pear-node
          </td>
          <td
            style="word-break: break-word;"
          >
            one, two
          </td>
          <td />
          <td>
            2019-05-22 00:00:51
          </td>
          <td>
            1682c475-04fd-5c4f-8881-7d4d44a3bfb1
          </td>
          <td
            align="right"
          >
            <a
              class="c14"
              href="/web/cluster/localhost/session/1682c475-04fd-5c4f-8881-7d4d44a3bfb1/player"
              kind="primary"
              target="_blank"
              width="80px"
            >
              Play
            </a>
          </td>
        </tr>
        <tr>
          <td>
            peach-node
          </td>
          <td
            style="word-break: break-word;"
          >
            one, two
          </td>
          <td />
          <td>
            2019-04-22 00:00:51
          </td>
          <td>
            377875-6491-11e9-80a1-427cfde50f5a
          </td>
          <td
            align="right"
          >
            <a
              class="c14"
              href="/web/cluster/localhost/session/377875-6491-11e9-80a1-427cfde50f5a/player"
              kind="primary"
              target="_blank"
              width="80px"
            >
              Play
            </a>
          </td>
        </tr>
        <tr>
          <td>
            apple-node
          </td>
          <td
            style="word-break: break-word;"
          >
            one, two
          </td>
          <td />
          <td>
            2019-04-22 00:00:51
          </td>
          <td>
            426485-6491-11e9-80a1-427cfde50f5a
          </td>
          <td
            align="right"
          >
            <a
              class="c14"
              href="/web/cluster/localhost/session/426485-6491-11e9-80a1-427cfde50f5a/player"
              kind="primary"
              target="_blank"
              width="80px"
            >
              Play
            </a>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`rendering of Session Recordings 2`] = `
.c14 {
  line-height: 1.5;
  margin: 0;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  box-sizing: border-box;
  border: none;
  border-radius: 4px;
  cursor: pointer;
  font-family: inherit;
  font-weight: 600;
  outline: none;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all 0.3s;
  -webkit-font-smoothing: antialiased;
  background: #512FC9;
  color: rgba(255,255,255,0.87);
  font-size: 10px;
  min-height: 24px;
  padding: 0px 16px;
  width: 80px;
}

.c14:active {
  opacity: 0.56;
}

.c14:hover,
.c14:focus {
  background: #651FFF;
}

.c14:active {
  background: #354AA4;
}

.c14:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c5 {
  display: inline-block;
  transition: color .3s;
  margin-left: 8px;
  margin-right: -8px;
  color: rgba(255,255,255,0.87);
  font-size: 14px;
}

.c11 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
  font-size: 16px;
}

.c13 {
  display: inline-block;
  transition: color .3s;
  color: #FFFFFF;
}

.c3 {
  line-height: 1.5;
  border-radius: 4px;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  border: 1px solid;
  box-sizing: border-box;
  background-color: transparent;
  cursor: pointer;
  font-family: inherit;
  font-weight: bold;
  outline: none;
  opacity: .56;
  position: relative;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  transition: all .3s;
  -webkit-font-smoothing: antialiased;
  border-color: rgba(255,255,255,0.87);
  color: #FFFFFF;
  font-size: 10px;
  min-height: 24px;
  padding: 0px 16px;
  margin-left: auto;
  width: auto;
  border-color: rgba(255,255,255,0.87);
  color: rgba(255,255,255,0.87);
}

.c3:hover {
  opacity: 1;
}

.c3:active {
  opacity: .24;
}

.c3 > span {
  display: flex;
  align-items: center;
  justify-content: center;
}

.c3:hover,
.c3:focus {
  border-color: #FFFFFF;
  color: #FFFFFF;
}

.c3:disabled {
  background: rgba(255,255,255,0.12);
  color: rgba(255,255,255,0.3);
}

.c3:hover,
.c3:focus {
  border-color: #FFFFFF;
  color: #FFFFFF;
}

.c9 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 400;
  font-size: 12px;
  line-height: 16px;
  margin: 0px;
  color: #FFFFFF;
}

.c6 {
  box-sizing: border-box;
  margin-bottom: 24px;
  flex: 0 0 auto;
  display: flex;
  align-items: center;
  justify-content: flex-start;
}

.c1 {
  box-sizing: border-box;
  margin-bottom: 24px;
  display: flex;
  align-items: center;
  flex-shrink: 0;
  border-bottom: 1px solid #1C254D;
  height: 56px;
  margin-left: -40px;
  margin-right: -40px;
  padding-left: 40px;
  padding-right: 40px;
}

.c2 {
  overflow: hidden;
  text-overflow: ellipsis;
  font-weight: 300;
  font-size: 22px;
  line-height: 32px;
  margin: 0px;
  margin-right: 56px;
  white-space: nowrap;
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  display: flex;
  overflow: auto;
  width: 100%;
  height: 100%;
  flex-direction: column;
}

.c0::after {
  content: ' ';
  padding-bottom: 24px;
}

.c12 {
  background: #222C59;
  box-shadow: 0 4px 16px rgba(0,0,0,0.24);
  border-collapse: collapse;
  border-spacing: 0;
  border-radius: 8px;
  font-size: 12px;
  width: 100%;
}

.c12 > thead > tr > th,
.c12 > tbody > tr > th,
.c12 > tfoot > tr > th,
.c12 > thead > tr > td,
.c12 > tbody > tr > td,
.c12 > tfoot > tr > td {
  padding: 8px 8px;
  vertical-align: middle;
}

.c12 > thead > tr > th:first-child,
.c12 > tbody > tr > th:first-child,
.c12 > tfoot > tr > th:first-child,
.c12 > thead > tr > td:first-child,
.c12 > tbody > tr > td:first-child,
.c12 > tfoot > tr > td:first-child {
  padding-left: 24px;
}

.c12 > thead > tr > th:last-child,
.c12 > tbody > tr > th:last-child,
.c12 > tfoot > tr > th:last-child,
.c12 > thead > tr > td:last-child,
.c12 > tbody > tr > td:last-child,
.c12 > tfoot > tr > td:last-child {
  padding-right: 24px;
}

.c12 > thead > tr > th {
  background: #111B48;
  color: #FFFFFF;
  cursor: pointer;
  font-size: 10px;
  font-weight: 400;
  padding-bottom: 0;
  padding-top: 0;
  text-align: left;
  opacity: 0.75;
  text-transform: uppercase;
  white-space: nowrap;
}

.c12 > thead > tr > th .c4 {
  font-weight: bold;
  font-size: 8px;
  margin-left: 8px;
}

.c12 > tbody > tr > td {
  color: rgba(255,255,255,0.87);
  line-height: 16px;
}

.c12 tbody tr {
  border-bottom: 1px solid #1C254D;
}

.c12 tbody tr:hover {
  background-color: rgb(37,49,98);
}

.c12 tbody tr:hover:last-child td:first-child {
  border-bottom-left-radius: 8px;
}

.c12 tbody tr:hover:last-child td:last-child {
  border-bottom-right-radius: 8px;
}

.c10 {
  box-sizing: border-box;
  display: flex;
}

.c10 button {
  background: none;
  border: none;
  border-radius: 200px;
  cursor: pointer;
  padding: 0;
  margin: 0 0 0 8px;
  outline: none;
  transition: all 0.3s;
  text-align: center;
}

.c10 button:hover,
.c10 button:focus {
  background: #1C254D;
}

.c10 button:hover .c4,
.c10 button:focus .c4 {
  opacity: 1;
}

.c10 button .c4 {
  opacity: 0.4;
  font-size: 20px;
  transition: all 0.3s;
}

.c8 {
  padding: 8px 24px;
  display: flex;
  height: 24px;
  flex-shrink: 0;
  align-items: center;
  justify-content: space-between;
  background: #222C59;
  border-top-left-radius: 8px;
  border-top-right-radius: 8px;
}

.c7 {
  box-sizing: border-box;
  font-size: 12px;
  min-width: 200px;
  outline: none;
  border: none;
  border-radius: 200px;
  height: 32px;
  transition: all .2s;
  background: #222C59;
  margin-right: 16px;
  padding-left: 16px;
  padding-right: 16px;
  color: rgba(255,255,255,0.87);
}

.c7:hover {
  background: #2C3A73;
}

.c7:focus,
.c7:active {
  background: #2C3A73;
  box-shadow: inset 0 2px 4px rgba(0,0,0,.24);
  color: rgba(255,255,255,0.87);
}

.c7::placeholder {
  color: rgba(255,255,255,0.24);
  font-size: 12px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
         Session Recordings
      </div>
      <button
        class="c3"
        style="min-width: 100px;"
        width="auto"
      >
        <span>
          Today
          <span
            class="c4 c5 icon icon-caret-down "
            color="text.primary"
            font-size="2"
          />
        </span>
      </button>
    </div>
    <div
      class="c6"
    >
      <input
        class="c7"
        color="text.primary"
        placeholder="SEARCH..."
        value=""
      />
    </div>
    <div>
      <nav
        class="c8"
      >
        <div
          class="c9"
          color="primary.contrastText"
        >
          SHOWING 
          <strong>
            1
          </strong>
           - 
          <strong>
            3
          </strong>
           
          of 
          <strong>
            3
          </strong>
        </div>
        <div
          class="c10"
        >
          <button
            disabled=""
            title="Previous Page"
          >
            <span
              class="c4 c11 icon icon-arrow-left-circle "
              color="light"
              font-size="3"
            />
          </button>
          <button
            disabled=""
            title="Next Page"
          >
            <span
              class="c4 c11 icon icon-arrow-right-circle "
              color="light"
              font-size="3"
            />
          </button>
        </div>
      </nav>
      <table
        class="c12"
      >
        <thead>
          <tr>
            <th>
              Node
            </th>
            <th>
              User(s)
            </th>
            <th>
              <a>
                Duration
                <span
                  class="c4 c13 icon icon-chevrons-expand-vertical "
                  color="light"
                />
              </a>
            </th>
            <th>
              <a>
                Created
                <span
                  class="c4 c13 icon icon-chevron-up "
                  color="light"
                />
              </a>
            </th>
            <th>
              Session ID
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              pear-node
            </td>
            <td
              style="word-break: break-word;"
            >
              one, two
            </td>
            <td />
            <td>
              2019-05-22 00:00:51
            </td>
            <td>
              1682c475-04fd-5c4f-8881-7d4d44a3bfb1
            </td>
            <td
              align="right"
            >
              <a
                class="c14"
                href="/web/cluster/localhost/session/1682c475-04fd-5c4f-8881-7d4d44a3bfb1/player"
                kind="primary"
                target="_blank"
                width="80px"
              >
                Play
              </a>
            </td>
          </tr>
          <tr>
            <td>
              peach-node
            </td>
            <td
              style="word-break: break-word;"
            >
              one, two
            </td>
            <td />
            <td>
              2019-04-22 00:00:51
            </td>
            <td>
              377875-6491-11e9-80a1-427cfde50f5a
            </td>
            <td
              align="right"
            >
              <a
                class="c14"
                href="/web/cluster/localhost/session/377875-6491-11e9-80a1-427cfde50f5a/player"
                kind="primary"
                target="_blank"
                width="80px"
              >
                Play
              </a>
            </td>
          </tr>
          <tr>
            <td>
              apple-node
            </td>
            <td
              style="word-break: break-word;"
            >
              one, two
            </td>
            <td />
            <td>
              2019-04-22 00:00:51
            </td>
            <td>
              426485-6491-11e9-80a1-427cfde50f5a
            </td>
            <td
              align="right"
            >
              <a
                class="c14"
                href="/web/cluster/localhost/session/426485-6491-11e9-80a1-427cfde50f5a/player"
                kind="primary"
                target="_blank"
                width="80px"
              >
                Play
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
